src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs

```diff
@@ -209,6 +209,38 @@ private WindowsIdentity(IntPtr userToken, string authType, int isAuthenticated)
             _isAuthenticated = isAuthenticated;
         }
 
+        private static SafeAccessTokenHandle DuplicateAccessToken(IntPtr accessToken)
+        {
+            SafeAccessTokenHandle duplicateAccessToken = SafeAccessTokenHandle.InvalidHandle;
+            IntPtr currentProcessHandle = Interop.Kernel32.GetCurrentProcess();
+            if (!Interop.Kernel32.DuplicateHandle(
+                    currentProcessHandle,
+                    accessToken,
+                    currentProcessHandle,
+                    ref duplicateAccessToken,
+                    0,
+                    true,
+                    Interop.DuplicateHandleOptions.DUPLICATE_SAME_ACCESS))
+            {
+                throw new SecurityException(new Win32Exception().Message);
+            }
+
+            return duplicateAccessToken;
+        }
+
+        private static SafeAccessTokenHandle DuplicateAccessToken(SafeAccessTokenHandle accessToken)
+        {
+            Debug.Assert(accessToken != null);
+
+            if (accessToken.IsInvalid)
+            {
+                return accessToken;
+            }
+
+            SafeAccessTokenHandle duplicateAccessToken = DuplicateAccessToken(accessToken.DangerousGetHandle());
+            GC.KeepAlive(accessToken);
```

Comment:
Should we do the whole DangerousAddRef/DangerousRelease dance instead?

Reply:
Could do that too, this is what is done elsewhere and I believe it suffices and is cheaper

Reply:
> this is what is done elsewhere

From what I can see, netfx passes the SafeHandle to the P/Invoke:
https://referencesource.microsoft.com/#mscorlib/system/security/principal/windowsimpersonationcontext.cs,47
and it will do the addref/release as part of that marshaling.

> I believe it suffices and is cheaper

I'm not convinced it's sufficient.  The KeepAlive prevents the access token from being finalized while it's being used in the P/Invoke, but it doesn't prevent Dispose from being called on the SafeHandle.  This SafeHandle was passed in by the developer, so concurrent with this operation, they could be calling Dispose on it.  If that Dispose happened after the IsInvalid checks, it could be closing the handle prior to or during the DuplicateHandle operation, which then also makes it susceptible to handle recycling.

Reply:
From the API perspective, a concurrent dispose is something the caller is responsible to prevent. That can also happen with the solution you suggest.

Reply:
I'm not following, if the disposal happens after the p/invoke, how it is better than the worst case scenario than the current state?

Reply:
> I'm not following, if the disposal happens after the p/invoke, how it is better than the worst case scenario of the current state?

The worst case is the dispose happens before or during the P/Invoke, at which point the P/Invoke could attempt to operate on a close handle, or worse, on a completely unrelated handle that was opened by Windows with the same value after the previous one was closed but for a completely unrelated resource.

Reply:
Also, Dispose is not enough to recycle a SafeHandle, hence the GC methods in the test case to actually cause the handle to be closed. From my understanding it needs to be finalized before the handle would actually be closed.

Reply:
> From my understanding it needs to be finalized before the handle would actually be closed.

This is not true.  If you call Dispose on a SafeHandle and it doesn't have a positive reference could from DangerousAddRef having been used (either directly or via marshaling), it will synchronously call ReleaseHandle to close the handle.

Reply:
If the dispose happens before the p/invoke (assuming you're talking about DuplicateHandle), the p/invoke should recognize that it's an invalid handle and fail, unless that handle value is reused. Ref counting does not solve that problem, as the dispose may happen after the SubRef (or equivalent), at which point the handle is invalid and any attempt to do anything meaningful on it later would fail in the same way. In-between the concurrent closing and reuse of the handle, the handle could be reused, in the same say you mentioned. I still don't see any significant difference.

> This is not true

Yes you are right, but you're talking about a SafeHandle with one ref that goes to zero refs with a concurrent dispose. How is that the API's problem?

Reply:
> The KeepAlive prevents the access token from being finalized while it's being used in the P/Invoke, but it doesn't prevent Dispose from being called on the SafeHandle

A concurrent dispose may happen at any time. An AddRef would prevent the handle from being closed, but only so long as the added ref is in effect; following the SubRef, all bets are off, just like GC.KeepAlive. There is some additional protection with ref counting but it's not warranted when it can face the same issues as the alternative.

Reply:
> From what I can see, netfx passes the SafeHandle to the P/Invoke

It also uses the GC.KeepAlive mechanism here:
https://referencesource.microsoft.com/#mscorlib/system/security/principal/windowsidentity.cs,128
Which is used in WindowsIdentity.RunImpersonated and a few other places.

I don't think there's a significant difference, but I can change it to ref counting if you prefer.

Reply:
> If the SafeHandle is disposed concurrently, with the solution I suggest it is guaranteed that either the dispose will happen prior to the P/Invoke and result in an exception being thrown then

The handle may also be recycled before the P/Invoke, in which case the P/Invoke may not result in an exception.

Reply:
I think we're talking past each other.

SafeHandle completely protects the underlying handle against recycling.  If it's disposed of, any subsequent attempts to use it will result in exception, e.g. passing a disposed SafeHandle to a P/Invoke will throw because the P/Invoke will call DangerousAddRef, which will use an interlocked to try to get a ref on the SafeHandle and will fail because it was already disposed.  And if it's currently being used in a P/Invoke, any disposals will be delayed until the P/Invoke completes, because the disposal will only take effect the moment the ref count drops to 0 when the P/Invoke marshaling layer invokes DangerousRelease.

I think this is potentially quite impactful here.  We're talking about security and impersonation.  Imagine this is in an ASP.NET web site, where every request is impersonating the caller, and that impersonation is then used to determine, for example, whether a particular file can be opened.  And let's say the code has a bug where the AccessToken is disposed of concurrently with this operation.  If a SafeHandle is used appropriately, either passed to a P/Invoke so that AddRef/Release are used or with the code using AddRef/Release directly, then there is no way the DuplicateHandle call could end up with a recycled handle: either disposal of the SafeHandle would happen prior to the AddRef, in which case AddRef will throw, or the disposal won't happen until after DuplicateHandle has completed.  But with an IntPtr and not using AddRef/Release, the disposal could free the handle, another unrelated request could create an AccessToken for a completely different user with different privileges and with the same handle value, and then this DuplicateHandle call could end up duplicating that other handle with greater privileges... then this request proceeds happily thinking it has a handle that represents one user when it reality it represents a completely different user.

Reply:
Responding to myself:
> Ref counting does not solve that problem, as the dispose may happen after the SubRef (or equivalent), at which point the handle is invalid and any attempt to do anything meaningful on it later would fail in the same way.

That's not quite right, if the dispose happens after the duplicate then it's fine because only the (still valid) duplicate would be used later. The issue is when the dispose happens before (either before the AddRef or before the actual duplication, between which there is no difference).

Reply:
> The issue is when the dispose happens before (either before the AddRef or before the actual duplication, between which there is no difference).

There is a difference.  If dispose happens before the AddRef, then AddRef will throw an exception.  If it happens after the AddRef, then the disposal will be delayed until the Release after the P/Invoke.  Either way, it's guaranteed that if the P/Invoke is called, it'll be on the right and open handle.

Reply:
I see, I missed that the implied AddRef would throw. That's a fair point, I'll change it.